pkg/webhook/cspc.go

```diff
@@ -0,0 +1,433 @@
+			cspc.Name,
+			repeatedBlockDevices)
+	}
+
+	buildPoolValidator := NewBuilder().
+		withPoolNamespace().
+		withCSPCName(cspc.Name).
+		withClientset(wh.clientset)
+	for _, pool := range cspc.Spec.Pools {
+		pool := pool // pin it
+		nodeName, err := GetNodeFromLabelSelector(pool.NodeSelector, wh.kubeClient)
+		if err != nil {
+			return false, fmt.Sprintf(
+				"failed to get node from pool nodeSelector: {%v} error: {%v}",
+				pool.NodeSelector,
+				err,
+			)
+		}
+		if usedNodes[nodeName] {
+			return false, fmt.Sprintf("invalid cspc: duplicate node %s entry", nodeName)
+		}
+		usedNodes[nodeName] = true
+		pValidate := buildPoolValidator.withPoolSpec(pool).
+			withPoolNodeName(nodeName).build()
+		ok, msg := pValidate.poolSpecValidation()
+		if !ok {
+			return false, fmt.Sprintf("invalid pool spec: %s", msg)
+		}
+	}
+	return true, ""
+}
+
+// getDuplicateBlockDeviceList returns list of block devices that are
+// duplicated in CSPC
+func getDuplicateBlockDeviceList(cspc *cstor.CStorPoolCluster) []string {
+	duplicateBlockDeviceList := []string{}
+	blockDeviceMap := map[string]bool{}
+	addedBlockDevices := map[string]bool{}
+	for _, poolSpec := range cspc.Spec.Pools {
+		for _, raidGroup := range poolSpec.DataRaidGroups {
+			for _, bd := range raidGroup.BlockDevices {
+				// update duplicateBlockDeviceList only if block device is
+				// repeated in CSPC and doesn't exist in duplicate block device
+				// list.
+				if blockDeviceMap[bd.BlockDeviceName] &&
+					!addedBlockDevices[bd.BlockDeviceName] {
+					duplicateBlockDeviceList = append(
+						duplicateBlockDeviceList,
+						bd.BlockDeviceName)
+					addedBlockDevices[bd.BlockDeviceName] = true
+				} else if !blockDeviceMap[bd.BlockDeviceName] {
+					blockDeviceMap[bd.BlockDeviceName] = true
+				}
+			}
+		}
+	}
+	return duplicateBlockDeviceList
+}
+
+func (poolValidator *PoolValidator) poolSpecValidation() (bool, string) {
+	if len(poolValidator.poolSpec.DataRaidGroups) == 0 {
+		return false, "at least one raid group should be present on pool spec"
+	}
+	// TODO : Add validation for pool config
+	// Pool config will require mutating webhooks also.
+	ok, msg := poolValidator.poolConfigValidation(poolValidator.poolSpec.PoolConfig)
+	if !ok {
+		return false, msg
+	}
+	for _, raidGroup := range poolValidator.poolSpec.DataRaidGroups {
+		raidGroup := raidGroup // pin it
+		ok, msg := poolValidator.raidGroupValidation(&raidGroup, poolValidator.poolSpec.PoolConfig.DataRaidGroupType)
+		if !ok {
+			return false, msg
+		}
+	}
+
+	return true, ""
+}
+
+var (
+	// SupportedPRaidType is a map holding the supported raid configurations
+	// Value of the keys --
+	// 1. In case of striped this is the minimum number of disk required.
+	// 2. In all other cases this is the exact number of disks required.
+	SupportedPRaidType = map[cstor.PoolType]int{
+		cstor.PoolStriped:  1,
+		cstor.PoolMirrored: 2,
+		cstor.PoolRaidz:    3,
+		cstor.PoolRaidz2:   6,
+	}
+)
```

Comment:
ToDo: Fix the min validation in next PR. As for raidz1 and raidz2 the min is 2^n+1 and 2^n+2 respectively,